osquery/core/watcher.cpp

```diff
@@ -345,21 +345,29 @@ bool WatcherRunner::watch(const PlatformProcess& child) const {
 }
 
 void WatcherRunner::stopChild(const PlatformProcess& child) const {
-  child.killGracefully();
+  const auto childPid = child.pid();
+  if (childPid < 0) {
+    const auto message = std::string("Watcher cannot stop worker process (") + std::to_string(childPid) + ").";
+    LOG(INFO) << message;
+    return;
+  }
 
-  // Clean up the defunct (zombie) process.
-  if (!child.cleanup()) {
-    auto child_pid = child.pid();
+  const auto killedGracefully = child.killGracefully();
+  if (killedGracefully) {
+    LOG(INFO) << "Pid: " << childPid << " killed gracefully";
+    return;
+  }
 
-    LOG(WARNING) << "osqueryd worker (" << std::to_string(child_pid)
-                 << ") could not be stopped. Sending kill signal.";
+  LOG(WARNING) << "osqueryd worker (" << childPid << ") could not be stopped. Sending kill signal.";
```

Comment:
You should add some kind of delay between the graceful kill, and the not graceful one.